### PR TITLE
lighting at the speed of light

### PR DESCRIPTION
**Pull request by Laboredih123**
[PERFORMANCE][INTERNAL]
## About the PR 
Optimises the lighting system a fair bit.
image
image
image

## Why's this needed? 
speed very good

### CODE REVIEW

File: code/_lighting/SimpleLight.dm [hunk 1]
```diff
@@ -346,44 +378,44 @@
...
 			vx = 0
-			vy = 1
+			vy = 32
 		if (NORTHEAST)
-			vx = 0.7071
-			vy = 0.7071
+			vx = sqrt(2)/16
+			vy = sqrt(2)/16
 		if (EAST)
 			vx = 1
```
Comment (Azrun) [inline code/_lighting/SimpleLight.dm:386]:
```suggestion
			vx = 32
```

File: code/_lighting/SimpleLight.dm [hunk 2]
```diff
@@ -346,44 +378,44 @@
...
+			vy = -sqrt(2)/16
 		if (WEST)
-			vx = -1
+			vx = -32
 			vy = 0
 		if (NORTHWEST)
-			vx = -0.7071
-			vy = 0.7071
+			vx = -sqrt(2)/16
+			vy = sqrt(2)/16
```
Comment (Azrun) [inline code/_lighting/SimpleLight.dm:402]: 1 -> `32`
0.7071 -> `sqrt(2)*16` ≈ `22.627`

File: code/_lighting/SimpleLight.dm [hunk 3]
```diff
@@ -249,12 +273,16 @@
 	show_mdir_light()
 
 	if (length(mdir_light_rgbas) == 1) //dont loop/average if list only contains 1 thing
-		for(var/obj/overlay/simple_light/medium/directional/mdir_light in src.mdir_lights)
-			if(mdir_light.dist == mdir_light_dists[mdir_light_dists.len])
-				mdir_light.color = rgb(rgba[1], rgba[2], rgba[3], min(255, rgba[4]))
+		var/alpha = rgba[4]
+		var/alpha_scaled = min(255, alpha / 2)
```
Comment (Azrun) [inline code/_lighting/SimpleLight.dm:277]: 0.5 != 0.4, is this just because it is cheaper to >>?

**Review by Azrun — CHANGES_REQUESTED** [review]

File: code/_lighting/SimpleLight.dm (hunk 3 above)
Comment (Laboredih123) [reply code/_lighting/SimpleLight.dm:277]: oops was copypasting

File: code/_lighting/SimpleLight.dm (hunk 2 above)
Comment (Laboredih123) [reply code/_lighting/SimpleLight.dm:402]: thank

File: code/_lighting/RobustLight2.dm [hunk 4]
```diff
@@ -735,20 +721,18 @@ proc/get_moving_lights_stats()
...
 
 /turf/proc/RL_Init()
-	if (!fullbright && !loc:force_fullbright)
-		if(!src.RL_MulOverlay)
-			src.RL_MulOverlay = new /obj/overlay/tile_effect/lighting/mul(src)
+	if (fullbright || loc:force_fullbright)
+		src.RL_MulOverlay?.dispose()
+		src.RL_MulOverlay = null
+		src.RL_AddOverlay?.dispose()
+		src.RL_AddOverlay = null
```
Comment (Azrun) [inline code/_lighting/RobustLight2.dm:728]: Provide note for why `dispose` is being used directly.

File: code/_lighting/SimpleLight.dm [hunk 5]
```diff
@@ -96,12 +112,14 @@
 		destroy_simple_light()
 
 
-/obj/overlay/simple_light/medium
+/atom/movable/light/simple_light/medium
 	icon_state = "medium_dir"
 	New(loc, dir=0)
 		..()
-		src.set_dir(dir)
+		src.dir = dir
```
Comment (Azrun) [inline code/_lighting/SimpleLight.dm:119]: Please add justification.

File: code/_lighting/SimpleLight.dm [hunk 6]
```diff
@@ -209,33 +231,35 @@
 		destroy_medium_light()
 
 /atom/proc/update_medium_light_visibility()
-	if(src.medium_lights[1].invisibility == 101) // toggled off
+	if(src.medium_lights[1].invisibility == INVIS_ALWAYS) // toggled off
```
Comment (Azrun) [inline code/_lighting/SimpleLight.dm:234]: ❤️

File: code/_lighting/SimpleLight.dm [hunk 7]
```diff
@@ -209,33 +231,35 @@
...
 		return
-	if(!isturf(src.loc) && !isturf(src))
-		for (var/obj/overlay/simple_light/medium/light as anything in src.medium_lights)
+	if(!(isturf(src.loc) || isturf(src)))
```
Comment (Azrun) [inline code/_lighting/SimpleLight.dm:236]: Won't both of these short circuit the same way? Not sure what the improvement is. 
Both conditions will be evaluated for the true path of the decision.

File: code/_lighting/SimpleLight.dm [hunk 8]
```diff
@@ -209,33 +231,35 @@
...
 		else
 			src:vis_contents += light
 
 
 
 
 
-/obj/overlay/simple_light/medium/directional
+
+
```
Comment (Azrun) [inline code/_lighting/SimpleLight.dm:255]: Why more whitespace?

File: code/_lighting/RobustLight2.dm (hunk 4 above)
Comment (Laboredih123) [reply code/_lighting/RobustLight2.dm:728]: qdel is a fair bit slow, and nothing should access the overlay except its own turf

File: code/_lighting/SimpleLight.dm (hunk 8 above)
Comment (Laboredih123) [reply code/_lighting/SimpleLight.dm:255]: woops

File: code/_lighting/SimpleLight.dm [hunk 9]
```diff
@@ -209,33 +231,30 @@
...
-			src:vis_contents -= light
+	if(src.medium_lights[1].invisibility == INVIS_ALWAYS) // toggled off
 		return
-	for (var/obj/overlay/simple_light/medium/light as anything in src.medium_lights)
-		if(light.icon_state == "medium_center")
-			src:vis_contents += light
-			continue
-		var/turf/T = get_step(get_turf(src), light.dir)
-		if(T?.opacity || T?.opaque_atom_count)
+	if(isturf(src.loc) || isturf(src))
```
Comment (Azrun) [inline code/_lighting/SimpleLight.dm:236]: You have now changed the logic. 
Was
        `if(!isturf(src.loc) && !isturf(src))`
Vs
`if(!(isturf(src.loc) || isturf(src)))`

**Review by Azrun — CHANGES_REQUESTED** [review]

File: code/_lighting/SimpleLight.dm (hunk 9 above)
Comment (Laboredih123) [reply code/_lighting/SimpleLight.dm:236]: after checking, the logic works out because i swapped the blocks too

**Review by Azrun — APPROVED** [review]

File: code/_lighting/SimpleLight.dm [hunk 10]
```diff
@@ -128,12 +144,16 @@
...
+		var/alpha_scaled = min(255, alpha / 2)
+		for(var/atom/movable/light/simple_light/medium/medium_light as anything in src.medium_lights)
+			if(medium_light.dir)
 				// divided by two because the directional sprites are brighter
-				medium_light.color = rgb(rgba[1], rgba[2], rgba[3], min(255, rgba[4] / 2))
+				rgba[4] = alpha_scaled
+				medium_light.color = rgba
+			else
+				rgba[4] = min(255, alpha)
+				medium_light.color = rgba
```
Comment (ZeWaka) [inline code/_lighting/SimpleLight.dm:156]:
```suggestion
				rgba[4] = alpha_scaled
			else
				rgba[4] = min(255, alpha)
			medium_light.color = rgba
```

Comment (ZeWaka) [reply code/_lighting/SimpleLight.dm:156]: just extract yo

**Review by ZeWaka — APPROVED** [review]

**Comment by TobleroneSwordfish** [comment]
I rate this PR 3/10 - brought down all three servers at once.

**Pull request merged**